File: scripts/generate-luts.mjs
import { mkdirSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const outDir = join(__dirname, "..", "assets", "luts");
const SIZE = 17;

function clamp(v) {
  return Math.min(1, Math.max(0, v));
}

function lerp(a, b, t) {
  return a + (b - a) * t;
}

function sat(r, g, b, amount) {
  const l = 0.2126 * r + 0.7152 * g + 0.0722 * b;
  return [lerp(l, r, amount), lerp(l, g, amount), lerp(l, b, amount)];
}

function contrast(r, g, b, amount) {
  return [
    lerp(0.5, r, amount),
    lerp(0.5, g, amount),
    lerp(0.5, b, amount),
  ];
}

function liftGammaGain(r, g, b, lift, gamma, gain) {
  const apply = (c, i) => clamp(Math.max(0, c * gain[i] + lift[i]) ** (1 / gamma[i]));
  return [apply(r, 0), apply(g, 1), apply(b, 2)];
}

function sCurve(x, amount) {
  const t = clamp(x);
  const curved = t * t * (3 - 2 * t);
  return lerp(t, curved, amount);
}

const recipes = {
  "kodak-2383": {
    title: "Kodak 2383",
    map([r, g, b]) {
      let [cr, cg, cb] = contrast(r, g, b, 1.18);
      cr = sCurve(cr, 0.35);
      cg = sCurve(cg, 0.32);
      cb = sCurve(cb, 0.38);
      return liftGammaGain(cr, cg, cb, [0.02, 0.03, 0.04], [1.05, 1.02, 0.96], [1.08, 1.02, 0.92]);
    },
  },
  "teal-orange": {
    title: "Teal Orange",
    map([r, g, b]) {
      const l = 0.2126 * r + 0.7152 * g + 0.0722 * b;
      let cr = r + (1 - l) * -0.08 + l * 0.12;
      let cg = g + (1 - l) * 0.02 + l * 0.04;
      let cb = b + (1 - l) * 0.1 + l * -0.1;
      [cr, cg, cb] = sat(cr, cg, cb, 1.12);
      [cr, cg, cb] = contrast(cr, cg, cb, 1.16);
      return [clamp(cr), clamp(cg), clamp(cb)];
    },
  },
  "bleach-bypass": {
    title: "Bleach Bypass",
    map([r, g, b]) {
      let [cr, cg, cb] = sat(r, g, b, 0.42);
      [cr, cg, cb] = contrast(cr, cg, cb, 1.38);
      return liftGammaGain(cr, cg, cb, [0.01, 0.02, 0.04], [0.95, 0.97, 1.02], [1.02, 1.04, 1.08]);
    },
  },
  "golden-hour": {
    title: "Golden Hour",
    map([r, g, b]) {
      let [cr, cg, cb] = liftGammaGain(r, g, b, [0.04, 0.02, 0.0], [1.08, 1.04, 0.98], [1.12, 1.04, 0.86]);
      [cr, cg, cb] = sat(cr, cg, cb, 1.08);
      return contrast(cr, cg, cb, 0.92).map(clamp);
    },
  },
  "fuji-eterna": {
    title: "Fuji Eterna",
    map([r, g, b]) {
      let [cr, cg, cb] = contrast(r, g, b, 0.88);
      [cr, cg, cb] = sat(cr, cg, cb, 0.86);
      return liftGammaGain(cr, cg, cb, [0.03, 0.04, 0.03], [1.04, 1.06, 1.02], [0.98, 1.04, 1.0]);
    },
  },
  "night-neon": {
    title: "Night Neon",
    map([r, g, b]) {
      let [cr, cg, cb] = contrast(r, g, b, 1.28);
      cr = cr * 1.08 + 0.04;
      cb = cb * 1.16 + 0.06;
      cg = cg * 0.9;
      [cr, cg, cb] = sat(cr, cg, cb, 1.25);
      return [clamp(cr * 0.92), clamp(cg * 0.86), clamp(cb)];
    },
  },
  "vintage-fade": {
    title: "Vintage Fade",
    map([r, g, b]) {
      let [cr, cg, cb] = contrast(r, g, b, 0.78);
      [cr, cg, cb] = sat(cr, cg, cb, 0.72);
      return liftGammaGain(cr, cg, cb, [0.08, 0.07, 0.04], [1.1, 1.06, 1.02], [1.04, 0.98, 0.88]);
    },
  },
  "arctic-cool": {
    title: "Arctic Cool",
    map([r, g, b]) {
      let [cr, cg, cb] = contrast(r, g, b, 1.14);
      [cr, cg, cb] = sat(cr, cg, cb, 0.9);
      return liftGammaGain(cr, cg, cb, [0.0, 0.02, 0.05], [0.98, 1.02, 1.08], [0.92, 1.02, 1.1]);
    },
  },
  "autumn-ember": {
    title: "Autumn Ember",
    map([r, g, b]) {
      let cr = r * 1.1 + 0.04;
      let cg = g * 0.96 + 0.02;
      let cb = b * 0.82;
      [cr, cg, cb] = sat(cr, cg, cb, 1.1);
      return contrast(cr, cg, cb, 1.08).map(clamp);
    },
  },
  "bw-classic": {
    title: "BW Classic",
    map([r, g, b]) {
      const y = 0.25 * r + 0.65 * g + 0.1 * b;
      const c = sCurve(y, 0.4);
      return [clamp(c * 1.02), clamp(c * 1.0), clamp(c * 0.96)];
    },
  },
  "portrait-soft": {
    title: "Portrait Soft",
    map([r, g, b]) {
      let [cr, cg, cb] = contrast(r, g, b, 0.86);
      [cr, cg, cb] = sat(cr, cg, cb, 0.92);
      return liftGammaGain(cr, cg, cb, [0.05, 0.03, 0.03], [1.06, 1.03, 1.02], [1.06, 1.0, 0.96]);
    },
  },
  "ocean-cyan": {
    title: "Ocean Cyan",
    map([r, g, b]) {
      let [cr, cg, cb] = liftGammaGain(r, g, b, [0.0, 0.03, 0.05], [1.0, 1.06, 1.08], [0.9, 1.06, 1.12]);
      [cr, cg, cb] = sat(cr, cg, cb, 1.06);
      return contrast(cr, cg, cb, 1.06).map(clamp);
    },
  },
};

function writeCube(id, title, map) {
  const lines = [
    `# ${title}`,
    `# Generated demo LUT for LOOKTABLE`,
    `TITLE "${title}"`,
    "DOMAIN_MIN 0.0 0.0 0.0",
    "DOMAIN_MAX 1.0 1.0 1.0",
    `LUT_3D_SIZE ${SIZE}`,
    "",
  ];

  for (let kb = 0; kb < SIZE; kb += 1) {
    for (let kg = 0; kg < SIZE; kg += 1) {
      for (let kr = 0; kr < SIZE; kr += 1) {
        const input = [kr / (SIZE - 1), kg / (SIZE - 1), kb / (SIZE - 1)];
        const [r, g, b] = map(input).map(clamp);
        lines.push(`${r.toFixed(6)} ${g.toFixed(6)} ${b.toFixed(6)}`);
      }
    }
  }

  writeFileSync(join(outDir, `${id}.cube`), `${lines.join("\n")}\n`);
}

mkdirSync(outDir, { recursive: true });
for (const [id, recipe] of Object.entries(recipes)) {
  writeCube(id, recipe.title, recipe.map);
  console.log(`wrote ${id}.cube`);
}

File: scripts/generate-luts.py
#!/usr/bin/env python3
from pathlib import Path

SIZE = 17
OUT_DIR = Path(__file__).resolve().parent.parent / "assets" / "luts"


def clamp(v):
    return min(1.0, max(0.0, v))


def lerp(a, b, t):
    return a + (b - a) * t


def sat(r, g, b, amount):
    luma = 0.2126 * r + 0.7152 * g + 0.0722 * b
    return lerp(luma, r, amount), lerp(luma, g, amount), lerp(luma, b, amount)


def contrast(r, g, b, amount):
    return lerp(0.5, r, amount), lerp(0.5, g, amount), lerp(0.5, b, amount)


def lift_gamma_gain(r, g, b, lift, gamma, gain):
    def apply(c, i):
        return clamp(max(0.0, c * gain[i] + lift[i]) ** (1 / gamma[i]))

    return apply(r, 0), apply(g, 1), apply(b, 2)


def s_curve(x, amount):
    t = clamp(x)
    curved = t * t * (3 - 2 * t)
    return lerp(t, curved, amount)


def kodak_2383(rgb):
    r, g, b = contrast(*rgb, 1.18)
    r, g, b = s_curve(r, 0.35), s_curve(g, 0.32), s_curve(b, 0.38)
    return lift_gamma_gain(r, g, b, (0.02, 0.03, 0.04), (1.05, 1.02, 0.96), (1.08, 1.02, 0.92))


def teal_orange(rgb):
    r, g, b = rgb
    luma = 0.2126 * r + 0.7152 * g + 0.0722 * b
    r = r + (1 - luma) * -0.08 + luma * 0.12
    g = g + (1 - luma) * 0.02 + luma * 0.04
    b = b + (1 - luma) * 0.10 + luma * -0.10
    r, g, b = sat(r, g, b, 1.12)
    r, g, b = contrast(r, g, b, 1.16)
    return clamp(r), clamp(g), clamp(b)


def bleach_bypass(rgb):
    r, g, b = sat(*rgb, 0.42)
    r, g, b = contrast(r, g, b, 1.38)
    return lift_gamma_gain(r, g, b, (0.01, 0.02, 0.04), (0.95, 0.97, 1.02), (1.02, 1.04, 1.08))


def golden_hour(rgb):
    r, g, b = lift_gamma_gain(*rgb, (0.04, 0.02, 0.0), (1.08, 1.04, 0.98), (1.12, 1.04, 0.86))
    r, g, b = sat(r, g, b, 1.08)
    return tuple(clamp(v) for v in contrast(r, g, b, 0.92))


def fuji_eterna(rgb):
    r, g, b = contrast(*rgb, 0.88)
    r, g, b = sat(r, g, b, 0.86)
    return lift_gamma_gain(r, g, b, (0.03, 0.04, 0.03), (1.04, 1.06, 1.02), (0.98, 1.04, 1.0))


def night_neon(rgb):
    r, g, b = contrast(*rgb, 1.28)
    r, g, b = r * 1.08 + 0.04, g * 0.9, b * 1.16 + 0.06
    r, g, b = sat(r, g, b, 1.25)
    return clamp(r * 0.92), clamp(g * 0.86), clamp(b)


def vintage_fade(rgb):
    r, g, b = contrast(*rgb, 0.78)
    r, g, b = sat(r, g, b, 0.72)
    return lift_gamma_gain(r, g, b, (0.08, 0.07, 0.04), (1.1, 1.06, 1.02), (1.04, 0.98, 0.88))


def arctic_cool(rgb):
    r, g, b = contrast(*rgb, 1.14)
    r, g, b = sat(r, g, b, 0.9)
    return lift_gamma_gain(r, g, b, (0.0, 0.02, 0.05), (0.98, 1.02, 1.08), (0.92, 1.02, 1.1))


def autumn_ember(rgb):
    r, g, b = rgb[0] * 1.1 + 0.04, rgb[1] * 0.96 + 0.02, rgb[2] * 0.82
    r, g, b = sat(r, g, b, 1.1)
    return tuple(clamp(v) for v in contrast(r, g, b, 1.08))


def bw_classic(rgb):
    y = 0.25 * rgb[0] + 0.65 * rgb[1] + 0.1 * rgb[2]
    c = s_curve(y, 0.4)
    return clamp(c * 1.02), clamp(c), clamp(c * 0.96)


def portrait_soft(rgb):
    r, g, b = contrast(*rgb, 0.86)
    r, g, b = sat(r, g, b, 0.92)
    return lift_gamma_gain(r, g, b, (0.05, 0.03, 0.03), (1.06, 1.03, 1.02), (1.06, 1.0, 0.96))


def ocean_cyan(rgb):
    r, g, b = lift_gamma_gain(*rgb, (0.0, 0.03, 0.05), (1.0, 1.06, 1.08), (0.9, 1.06, 1.12))
    r, g, b = sat(r, g, b, 1.06)
    return tuple(clamp(v) for v in contrast(r, g, b, 1.06))


RECIPES = {
    "kodak-2383": ("Kodak 2383", kodak_2383),
    "teal-orange": ("Teal Orange", teal_orange),
    "bleach-bypass": ("Bleach Bypass", bleach_bypass),
    "golden-hour": ("Golden Hour", golden_hour),
    "fuji-eterna": ("Fuji Eterna", fuji_eterna),
    "night-neon": ("Night Neon", night_neon),
    "vintage-fade": ("Vintage Fade", vintage_fade),
    "arctic-cool": ("Arctic Cool", arctic_cool),
    "autumn-ember": ("Autumn Ember", autumn_ember),
    "bw-classic": ("BW Classic", bw_classic),
    "portrait-soft": ("Portrait Soft", portrait_soft),
    "ocean-cyan": ("Ocean Cyan", ocean_cyan),
}


def write_cube(lut_id, title, mapper):
    lines = [
        f"# {title}",
        "# Generated demo LUT for LOOKTABLE",
        f'TITLE "{title}"',
        "DOMAIN_MIN 0.0 0.0 0.0",
        "DOMAIN_MAX 1.0 1.0 1.0",
        f"LUT_3D_SIZE {SIZE}",
        "",
    ]
    last = SIZE - 1
    for kb in range(SIZE):
        for kg in range(SIZE):
            for kr in range(SIZE):
                mapped = mapper((kr / last, kg / last, kb / last))
                r, g, b = (clamp(v) for v in mapped)
                lines.append(f"{r:.6f} {g:.6f} {b:.6f}")
    (OUT_DIR / f"{lut_id}.cube").write_text("\n".join(lines) + "\n", encoding="utf-8")


def main():
    OUT_DIR.mkdir(parents=True, exist_ok=True)
    for lut_id, (title, mapper) in RECIPES.items():
        write_cube(lut_id, title, mapper)
        print(f"wrote {lut_id}.cube")


if __name__ == "__main__":
    main()
